Scripting des objets et des NPC

fmkr-project opened the pull request:
Fix #30 

Permet le codage de petits scripts pour automatiser certaines actions du jeu
Cf. manuel pour voir comment implémenter correctement un script.

fmkr-project committed:
Début d'implémentat° des scripts

Encore non fonctionnels

fmkr-project committed:
Update npc.py

Fonction non utilisée (et inconnue pour ma part) dans les imports.

fmkr-project committed:
Commande d'obtention d'objets

+ début de la commande de mouvement

fmkr-project committed:
Script de mouvement

Note : voir pourquoi le script est pour l'instant à usage unique

fmkr-project committed:
Scripts réutilisables, opérations logiques

Essayer de parler au NPC de gauche sans, puis avec un objet "thermo" dans son Sac.
Le NPC en bas à droite utilise les fonctions de génération aléatoire.

fmkr-project committed:
Passage des sons dans leur propre classe

+ Ajout de fonctions de script relatives au son
- Spam dans la console lorsqu'un son est joué, mais ce spam est pour le moment inévitable
En attente de l'implémentation du gameplay principal avant de continuer.

fmkr-project committed:
Scripts + lisibles, ébauche des flags des NPCs

fmkr-project committed:
Un peu plus de mapping

+ Ajout du dortoir L(3)
+ Certaines cartes élargies pour éviter les effets de bord qui se produisent si la map est trop petite (jeu pas centré sur le perso)
+ Correction de détails graphiques

fmkr-project committed:
Suppression d'objets du Sac

+ Ajout d'une fonction pour supprimer un objet.
+ Ajout d'un NPC en I104 qui démontre la fonction. Parler au NPC en bas à droite pour obtenir des objets.
+ Support de la suppression de tous les objets d'un type

fmkr-project committed:
Ajout des infobox

+ Infobox : une boîte de dialogue mais sans interlocuteur
+ NPC démontrant l'usage de la fonction de script associée
- Fix appel de la fonction de son
- Les infobox sont encore contenus dans la classe Dialogue, ils sont appelés avec un dialogue fantôme et un NPC tout aussi inexistant. Une seule boîte affichable pour l'instant

fmkr-project committed:
Infobox à plusieurs phrases & manuel

+ Support de l'écriture de plusieurs phrases pour une seule infobox à l'aide d'une nouvelle fonction
+ Ajout d'un pdf pour que tout le monde puisse apprendre à écrire ses scripts en fonction de ses idées

fmkr-project committed:
Scripts de salles

+ Possibilité de déclencher un script en entrant dans une map
- Bug lors du reset de la sauvegarde

fmkr-project committed:
Appels de scripts dans des scripts

+ Possibilité d'appeler des scripts à l'intérieur de scripts par la modification de la manière de traiter le script courant par le jeu
+ Ajout d'un script complet pour Skler et modification de sa position (il est maintenant dans la Cave)
+ Ajout de la I306

○ NOTE : J'ai changé la vitesse de sprint du joueur à titre de debug uniquement, on remettra la valeur d'origine (1.75) en release 0.3

- BUG : Dans le script denis_brogniart, enlever sfx provoque un plantage (impossibilité d'afficher le dialogue)

fmkr-project committed:
Opérations sur l'acc numérique

+ Ajout de fonctions pour effectuer les opérations de base sur l'acc num (+-*/)
+ Correction du bug denis_brogniart
+ Mise à jour du menu de debug afin d'afficher les variables relatives aux scripts
+ PNJ de test en I109
+ Rework de la texture de plancher
+ Ajout des salles K201 et L201

fmkr-project committed:
Modifications graphiques des NPC

+ Possibilité d'ajouter des NPC invisibles
+ Ajout du panneau devant l'escalier I
+ Ajout des salles I110 et L106 (Vie Sco)

fmkr-project committed:
Petit nettoyage et cartes

+ Suppression du spam dans la console
+ Ajout de quelques couloirs et de portes (non fonctionnelles)

fmkr-project committed:
Map rework, partie I

+ Modifications du tileset
+ Poursuite de la conventionnalisation des tailles des salles (pour éviter les effets de bord)
+ Petite modif d'un panneau
+ Ajout des salles suivantes : K103, K105, M101, M103, "O101" (Chapelle, située dans le bât O)

fmkr-project committed:
Manuel version finale

+ Introduction de quelques nouvelles fonctions

○ TODO : faire des boucles for ?

fmkr-project committed:
Passage des lignes de texte en YAML

+ Ajout du dossier locale regroupant les lignes de texte
+ Transfert de toutes les lignes de texte de la BDD vers le yaml afin de faciliter leur édition
+ une musique

fmkr-project committed:
Map rework, partie II

+ Ajout des couloirs M3 et N3, ainsi que des escaliers associés
+ Ajout de la N305 (à des fins de démonstration seulement)
+ Possibilité d'ajouter des chaînes de caractères du fichier locale et des chaînes codées en dur en même temps dans les infobox
+ Possibilité de déplacer le joueur en arrière-plan (utile pour les portes)
+ Refonte du tileset pour que le joueur puisse disparaître derrière lors des transitions (aussi utile pour les portes)
+ Du son

fmkr-project committed:
Début de transition sonore

+ La musique n'est plus coupée si la map d'arrivée possède le même fond sonore

fmkr-project committed:
Map rework, partie III

+ Ajout des bâtiments N0, P2, P3
+ Ajout d'une ébauche de Cour d'Honneur

fmkr-project committed:
Update src/sound.py

Co-authored-by: code-review-doctor[bot] <72320148+code-review-doctor[bot]@users.noreply.github.com>

Ordralfabetix reviewed (changes requested):
```pygame 2.1.2 (SDL 2.0.16, Python 3.9.2)
Hello from the pygame community. https://www.pygame.org/contribute.html
Traceback (most recent call last):
  File "/home/mateo/prepa-simulator/src/main.py", line 18, in <module>
    game = g.Game()
  File "/home/mateo/prepa-simulator/src/game.py", line 82, in __init__
    self.map_manager = maps.MapManager(self.screen, self)
  File "/home/mateo/prepa-simulator/src/maps.py", line 29, in __init__
    self.load_map(config["map_id"]) # Charge la carte où est le joueur
  File "/home/mateo/prepa-simulator/src/maps.py", line 34, in load_map
    self.map_name = self.game.game_data_db.execute("select file from maps where id = ?;", (self.map_id,)).fetchall()[0][0]
IndexError: list index out of range
```

fmkr-project commented:
> `pygame 2.1.2 (SDL 2.0.16, Python 3.9.2) Hello from the pygame community. https://www.pygame.org/contribute.html Traceback (most recent call last): File "/home/mateo/prepa-simulator/src/main.py", line 18, in <module> game = g.Game() File "/home/mateo/prepa-simulator/src/game.py", line 82, in __init__ self.map_manager = maps.MapManager(self.screen, self) File "/home/mateo/prepa-simulator/src/maps.py", line 29, in __init__ self.load_map(config["map_id"]) # Charge la carte où est le joueur File "/home/mateo/prepa-simulator/src/maps.py", line 34, in load_map self.map_name = self.game.game_data_db.execute("select file from maps where id = ?;", (self.map_id,)).fetchall()[0][0] IndexError: list index out of range `

Il faut reset le ficheir de sauvegarde

Ordralfabetix commented:
Toutes les touches autres que directionnelles font fermer le jeu.
```Traceback (most recent call last):
  File "/home/mateo/prepa-simulator/src/main.py", line 19, in <module>
    game.run()
  File "/home/mateo/prepa-simulator/src/game.py", line 141, in run
    inputs.handle_key_down_event(self, event) # Gestion des touches préssées
  File "/home/mateo/prepa-simulator/src/inputs.py", line 45, in handle_key_down_event
    elif event.key == controls["DEBUG"]:
KeyError: 'DEBUG'```

antoinech2 commented:
> Toutes les touches autres que directionnelles font fermer le jeu.
> 
> ```
>   File "/home/mateo/prepa-simulator/src/main.py", line 19, in <module>
>     game.run()
>   File "/home/mateo/prepa-simulator/src/game.py", line 141, in run
>     inputs.handle_key_down_event(self, event) # Gestion des touches préssées
>   File "/home/mateo/prepa-simulator/src/inputs.py", line 45, in handle_key_down_event
>     elif event.key == controls["DEBUG"]:
> KeyError: 'DEBUG'```
> ```

De même, réinitialiser le dossier de sauvegarde des paramètres `/stg`

fmkr-project committed:
Map rework, partie IV

+ Ajout des salles M202 (anglais) et N306 (pour l'instant, pas d'îlots)
+ Agrandissement des maps pour supprimer les effets de bord (maps I10x en particulier)
+ Légère augmentation du volume
+ Correction d'un bug sonore de trk14

fmkr-project committed:
Flags des maps / Map rework, partie V

+ Agrandissement des dernières salles subissant encore les effets de bords (L101 et Cave)
+ Mise en place d'un système de flags pour chaque map avec sauvegarde
+ Fonctions associées aux flags : obtenir la valeur d'un flag, le baisser ou le lever (un exemple est disponible en haut à gauche de la I104, parler aussi au nouveau PNJ de la L101)
+ Ajout d'une fonction pour interrompre un script
+ Ajout d'un script sans contenu
+ Résolution du bug du reset de sauvegarde (commit 6ee7f17477aaf2561b9a8ce0d72b7eb0cc082be4)

fmkr-project committed:
Mise à jour du manuel

+ Prise en charge par le manuel des nouvelles fonctions sur les flags des salles
+ Petite modification pour éviter la confusion entre flags des salles et flags des PNJ

fmkr-project committed:
Nouveaux flags des PNJ

+ Conventionnalisation du nom des fonctions de lecture / écriture des flags
+ Sauvegarde de l'état des flags des PNJ
En ce qui concerne le bug du commit précédent, il ne faut pas ouvrir la BDD de la save en même temps que le jeu...
On se rapproche d'une vraie interaction entre les différentes composantes du jeu

fmkr-project committed:
Amélioration de l'obtention d'objets

+ Infobox et son lors de l'obtention d'un objet
+ Les objets ne peuvent désormais être ramassés qu'une seule fois

fmkr-project committed:
Labels et ébauche de mission

+ Introduction des labels et des goto issus du BASIC, PNJ de test vers l'escalier de la CdH
+ Petites avancées sur #5
+ Renommage de movingscript en move pour rendre les scripts plus compacts et lisibles
+ Possibilité de supprimer tous les objets d'un type donné (utile pour les objets rares)
+ Exemple de mission : câble HDMI pour le PC de la H009 (le câble est caché sous la table du prof de la I108)

- Suppression du fichier objects.yaml inutilisé

fmkr-project committed:
Mise à jour du manuel

+ Ajout de label et goto
Note : label ne prend pas d'argument dans le fichier script, mais cette même commande a un argument une fois placée dans un script (ex : label('monLabel') ).
+ Ajout d'un exemple plus explicite, plus commenté, et plus utile

fmkr-project committed:
Boîtes à choix multiples

+ Ajout des boîtes à choix multiples / choicebox (pour l'instant, il n'y a que des boîtes Oui/Non)
+ Ajout des fonctions de script associées. Exemple en I109
+ Moins de spam dans la console

- TODO relativement important : nettoyer de fond en comble le fichier menu afin de rendre le code (beaucoup) plus lisible

fmkr-project committed:
Menu et script de sauvegarde

+ Ajout d'une commande pour sauvegarder le jeu
+ Ajout d'un script "standard" save qui demande au joueur s'il veut sauvegarder la partie
+ Ajout de fonctions permettant de modifier directement l'accumulateur booléen
+ Sous-menu de sauvegarde fonctionnel
+ PNJ de test en I109

fmkr-project committed:
Bugfix : traitement de plusieurs PNJ

+ Le jeu ne plante plus lorsqu'on parle à un PNJ qui n'a pas d'ID de dialogue commun avec le PNJ auquel le joueur a parlé auparavant

fmkr-project committed:
Guide de style

+ Ajout d'un guide de style pour conventionnaliser les scripts
+ Petit développement du script hdmi en ajoutant une choicebox
+ Ajout de deux textures d'objets

fmkr-project committed:
Modifs graphiques

+ Ajout d'un input lock pour bloquer le clavier lorsqu'un script est en cours d'exécution
+ Ajout de plusieurs textures de joueur (M1, M3 et MP)

fmkr-project committed:
Flags généraux

+ Ajout de flags généraux qui affectent toute la partie, pas seulement une salle ou un PNJ en particulier
+ Création de la table events dans la sauvegarde pour sauvegarder l'état de ces drapeaux
+ Démonstration en I109... attention, le choix est définitif !

fmkr-project committed:
Correction math, pseudo-boucle for

+ La fonction math fonctionne désormais correctement avec les opérateurs de base
+ Ajout d'un script démontrant la possibilité de créer des boucles for simples (démo en L101)

fmkr-project committed:
Petits changements graphiques

+ La Chapelle est maintenant dans son agencement actuel (alias agencement type Covid)
